### PR TITLE
Generic http/httpAsync DSL

### PR DESCRIPTION
**Pull request by Hakky54**
This is a pull request to enable dynamic http request, see issue mentioned here: https://github.com/rybalkinsd/kohttp/issues/169

Summary:
With the current version the end-user needs to explicitly specify which http context to use for each http method type, see here code snippet from the issue https://github.com/rybalkinsd/kohttp/issues/169:
```
package it.skrape.core.fetcher

import io.github.rybalkinsd.kohttp.client.defaultHttpClient
import io.github.rybalkinsd.kohttp.client.fork
import io.github.rybalkinsd.kohttp.dsl.*
import io.github.rybalkinsd.kohttp.dsl.context.HttpContext
import io.github.rybalkinsd.kohttp.ext.url
import it.skrape.core.Request
import it.skrape.core.Result
import org.jsoup.Connection

class KoFetcher(private val request: Request): Fetcher {
	private val client = defaultHttpClient.fork {
		followRedirects = request.followRedirects
		readTimeout = request.timeout.toLong()
	}

	override fun fetch(): Result {
		val requester = when(request.method) {
			Connection.Method.GET -> httpGet(client, getContext())
			Connection.Method.POST -> httpPost(client, getContext())
			Connection.Method.PUT -> httpPut(client, getContext())
			Connection.Method.DELETE -> httpDelete(client, getContext())
			Connection.Method.PATCH -> httpPatch(client, getContext())
			Connection.Method.HEAD -> httpHead(client, getContext())
			else -> throw UnsupportedOperationException("Method is not supported by KoHttp")
		}
		requester.use {
			return Result(
				responseBody = it.body()?.string() ?: "",
				statusCode = it.code(),
				statusMessage = it.message(),
				contentType = it.header("Content-Type"),
				headers = it.headers().names().associateBy({item -> item}, {item -> it.header(item, "")!!}),
				request = request
			)
		}
	}

	private fun getContext(): HttpContext.() -> Unit = {
		url(request.url)
		header {
			request.headers
			"User-Agent" to request.userAgent
			cookie {
				request.cookies
			}
		}
	}
} 
```

Provided solution
Within this pull request a http request can be executed by either specifying the http method or http context. The underlying function will determine during runtime which http context to use by the provided arguments. 

The above snippet can be rewritten with the following snippet:
```
package it.skrape.core.fetcher

import io.github.rybalkinsd.kohttp.client.defaultHttpClient
import io.github.rybalkinsd.kohttp.client.fork
import io.github.rybalkinsd.kohttp.dsl.*
import io.github.rybalkinsd.kohttp.dsl.context.HttpContext
import io.github.rybalkinsd.kohttp.ext.url
import it.skrape.core.Request
import it.skrape.core.Result
import org.jsoup.Connection

class KoFetcher(private val request: Request): Fetcher {
	private val client = defaultHttpClient.fork {
		followRedirects = request.followRedirects
		readTimeout = request.timeout.toLong()
	}

	override fun fetch(): Result {
	    method: Method = Method.valueOf(request.method.toString)
	    val requester = http(method = method, init = getContext)

		requester.use {
			return Result(
				responseBody = it.body()?.string() ?: "",
				statusCode = it.code(),
				statusMessage = it.message(),
				contentType = it.header("Content-Type"),
				headers = it.headers().names().associateBy({item -> item}, {item -> it.header(item, "")!!}),
				request = request
			)
		}
	}

	private fun getContext(): HttpContext.() -> Unit = {
		url(request.url)
		header {
			request.headers
			"User-Agent" to request.userAgent
			cookie {
				request.cookies
			}
		}
	}
}
```

**Note**
This pull request is currently WIP to get review input on the initial solution. Some unit tests are already added to provide example usages.

### CODE REVIEW

**Comment by Hakky54** [comment]
@rybalkinsd This feature was in my opinion quite challenging. It works but I don't think it is the most clever solution. What do you think? 

I still need to add all the unit test and do some additional clean-up.

**Comment by rybalkinsd** [comment]
@Hakky54 

Wow! I also expected it to be more challenging. 

I like the API

    http(method = ... ) {
        contextAction()
    }

A couple of things to think about:
1. Method and Context are strongly connected. Is there a way to connect them as Types?
`fun <T> http(..., method: Method<T>, init: T.() -> Unit) ...`
It will probably remove a declaration duality
2. We recently discussed with @IVSivak about integration with converters/deserialisation #176 
Both tasks could affect our high-level DSL, and it's better to have a coordinated view here. One of the options we discussed:
    
    val r: ResultType =  httpGet<ResultTypeType> {  }    

@IVSivak could you please share your thoughts?

**Comment by Hakky54** [comment]
Sure, I will check if it is possible to combine the method and the context with generics.

By the way I have the feeling that HttpContextUtil is a bit over complex, it is in my opinion not very readable. What do you guys think? Do you think there is a better way with plain kotlin to have something like that working?

**Comment by rybalkinsd** [comment]
> By the way I have the feeling that HttpContextUtil is a bit over complex, it is in my opinion not very readable. What do you guys think? Do you think there is a better way with plain kotlin to have something like that working?

It looks like a bit complicated, imo. But yeah, as I mentioned, maybe it's possible to resolve this using types.

Another option to consider is to make `Method` a sort of context factory.

**Comment by Hakky54** [comment]
I refactored a bit and the resulting code is simplified compared to the initial commit. Functionality wise it is exact the same as the previous commit, the only difference is that with this commit a http method is required when calling the http function.

The flows will be:
- http function can be called with Method.GET and HttpContext
- http function can be called with Method.GET and HttpGetContext
- http function will throw class cast exception when calling with Method.GET and HttpPostContext

- http function can be called with Method.POST and HttpContext
- http function can be called with Method.POST and HttpPostContext
- http function will throw class cast exception when calling with Method.POST and HttpGetContext

- http function can be called with Method.DELETE and HttpContext
- http function can be called with Method.DELETE and HttpDeleteContext
- http function will throw class cast exception when calling with Method.DELETE and HttpGetContext

I also wanted to check the type of `init: T.() -> Unit` within the when statement, but couldn't find a good way. So what I wanted to do is for example is to create only a HttpGetContext when the method is GET and the init variable is instance of `HttpGetContext.() -> Unit` or is instance of `HttpContext.() -> Unit` or else throw a specific runtime exception as it does not make sense to pass a HttpPostContext when creating a GET request. Another option could be that it is still allowed to pass a HttpPostContext when creating a GET request but behind the scenes we create a new HttpContext instance from the HttpPostContext and supply it to the HttpGetContext. In that way it will be save to pass any kind of sub class of HttpContext to the http function.

What do you think of the code change and of the flows? Is it ok to throw a class cast exception when someone is being funny by passing a HttpPostContext while executing a Get request?

**Comment by rybalkinsd** [comment]
I like the simplified version!

I was thinking about when it's possible to use exactly same request with any request method.
And it looks like strict typing is probably the thing we're looking for

```
fun http(client: Call.Factory = defaultHttpClient, method: Method, init: HttpContext.() -> Unit): Response {
    return client.newCall(method.makeContext().apply(init).makeRequest()).execute()
}
```
It will only allow to use the context initialisers applicable to any context.
It's possible to make an extension function `makeContext` based on `createHttpContext`
It could look like
```
internal fun Method.makeContext(): HttpContext {
    return when (this) {
        Method.GET -> HttpGetContext()
        Method.POST -> HttpPostContext()
        ...
    }
}
```

Another opportunity here is to support methods with body (inherited from `HttpPostContext`)

**Comment by Hakky54** [comment]
It looks like I need to learn a lot about Kotlin, it is quite different compared to Java. I didn't know that an existing enum could be enriched with a function outside the enum declaration. Awesome!

I applied your feedback, but when I remove the generic type `<T : HttpContext>` the test class fails to compile. It doesn't accept that I provide HttpPostContext to a function which accepts `HttpContext.() -> Unit` In my opinion it doesn't make sense because I am sending an instance of a child class of a HttpContext and the function parameter should be saying that it should accept HttpContext and all child classes of HttpContext. Or do you think I am missing something?
I committed the changes up till now.

Your other option regarding supporting method with body by inheritance from HttpPostContext gives me the idea that it is the best option to move from here on, but looking at OOP aren't we misusing HttpPostContext? What do you think of moving the body property to HttpContext (the base class) and disable it at the child classes which don't have a body, such as the HttpGetContext.

**Comment by Hakky54** [comment]
I Added both cases mentioned within my precious comment as two separate commits:

Cases:
1. Extending Method with a function: https://github.com/rybalkinsd/kohttp/pull/189/commits/f38308803a744c3718b702726700f284a67614ad
2. Move body to base class: https://github.com/rybalkinsd/kohttp/pull/189/commits/92cd43af92549b99b8318bd4c32d0f18d893d8ba

I am curious what you think of the both cases. As a mentioned earlier, I couldn't remove the generic type within the http function as it fail to compile wen passing a sub class of HttpContext

**Comment by rybalkinsd** [comment]
@Hakky54 
> 2. Move body to base class and disable it for GET and HEAD requests: 92cd43a
We tried it once in a while. Users are getting frustrated :) So I would try to avoid it.

> 1. Extending Method with a function: f383088
It looks good for me!
Just want to make sure what is the reason of `T` in 
```
internal fun <T : HttpContext> Method.createHttpContext(): T
```

> the test class fails to compile
Maybe I'm missing something, but
I also expected test class not to compile and It's valid imo

        http(method = Method.DELETE, init = fun HttpContext.() {
            host = "postman-echo.com"
            path = "/delete"
            // body should not compile
            body { }
        }


Imagine our use case: we have an identical request and we want time to time use POST or GET to for it.

I would go with a code like
```
    val ctx = {
        host = "..."
        port = "..."
        param { }
    }

   http(method = variableMethod, ctx)
```
In this case the only possible `ctx` type that will make sense for both cases is `HttpContext.() -> Unit`

File: kohttp-test/src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/HttpDslKtTest.kt [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                        "ccc" to 42
+                    }
+                }
+
+                param {
+                    "text" to "iphone"
+                    "lr" to 213
+                }
+
+                body {
```
Comment (rybalkinsd) [inline kohttp-test/src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/HttpDslKtTest.kt:151]: As I mentioned in c comment section, we should limit our support (at least in our first release for this feature)
So let's use only `HttpContext.() -> Unit`

Comment (Hakky54) [reply kohttp-test/src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/HttpDslKtTest.kt]: Yes, sounds good. I changed it back to only support HttpContext

**Comment by Hakky54** [comment]
Thanks again for the feedback! 

> Just want to make sure what is the reason of T in
>
> internal fun <T : HttpContext> Method.createHttpContext(): T

The reason was to support child classes of HttpContext such as HttPostContext and HttpGetContext. I removed the generic type T as you already mentioned that it is fine for the initial version to only support the base class HttpContext. I dropped case number 2 and refactored it. I have tried to finalize it. Added unit tests and updated the documentation. Let me know if you are ok with the changes.

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpAsyncDsl.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *  val response: Deferred<Response> = httpAsync(customHttpClient, Method.GET) {
+ *      ...
+ *  }
+ * </pre>
+ *
+ * @see Response
+ * @see HttpContext
+ * @see Method
+ *
+ * @since 0.11.2
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpAsyncDsl.kt:43]: Let's put `0.12.0` for all new API features!

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpAsyncDsl.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * @author hakky54
+ */
+fun httpAsync(
+        client: Call.Factory = defaultHttpClient,
+        method: Method,
+        init: HttpContext.() -> Unit
+): Deferred<Response> =
+        GlobalScope.async(context = Dispatchers.Unconfined) {
+            client.suspendCall(method.makeHttpContext().apply(init).makeRequest())
+        }
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpAsyncDsl.kt:53]:
```suggestion
        }
        
```

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpContext.kt [hunk 4]
```diff
@@ -94,13 +94,11 @@ enum class Method {
...
 internal interface IHttpContext {
     fun param(init: ParamContext.() -> Unit)
     fun header(init: HeaderContext.() -> Unit)
     fun makeRequest(): Request
 }
 
-
 @DslMarker
-annotation class HttpDslMarker
+annotation class HttpDslMarker
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpContext.kt:104]:
```suggestion
annotation class HttpDslMarker

```

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/httpDsl.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * @see Response
+ * @see HttpContext
+ * @see Method
+ *
+ * @since 0.11.2
+ * @author hakky54
+ */
+fun http(client: Call.Factory = defaultHttpClient, method: Method, init: HttpContext.() -> Unit): Response {
+    return client.newCall(method.makeHttpContext().apply(init).makeRequest()).execute()
+}
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/httpDsl.kt:49]:
```suggestion
}

```

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/httpDsl.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * </pre>
+ *
+ * @see Response
+ * @see HttpContext
+ * @see Method
+ *
+ * @since 0.11.2
+ * @author hakky54
+ */
+fun http(client: Call.Factory = defaultHttpClient, method: Method, init: HttpContext.() -> Unit): Response {
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/httpDsl.kt:47]: Let's use a new line for each parameter, as in `httpAsync`

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/util/HttpContextUtil.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+internal fun Method.makeHttpContext(): HttpContext {
+    return when (this) {
+        Method.GET -> HttpGetContext()
+        Method.POST -> HttpPostContext()
+        Method.PUT -> HttpPutContext()
+        Method.DELETE -> HttpDeleteContext()
+        Method.PATCH -> HttpPatchContext()
+        Method.HEAD -> HttpHeadContext()
+    }
+}
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/util/HttpContextUtil.kt:14]:
```suggestion
}

```

File: kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/util/HttpContextUtil.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package io.github.rybalkinsd.kohttp.util
+
+import io.github.rybalkinsd.kohttp.dsl.context.*
+
+internal fun Method.makeHttpContext(): HttpContext {
+    return when (this) {
+        Method.GET -> HttpGetContext()
```
Comment (rybalkinsd) [inline kohttp/src/main/kotlin/io/github/rybalkinsd/kohttp/util/HttpContextUtil.kt:7]: Let's import Method to use them like `Post -> HttpPostContext()`

**Comment by rybalkinsd** [comment]
> By the way I also added unit tests for all the async http requests, but somehow CodeCov isn't recognises it

No worries at all, I'll check codecov later this week

**Comment by rybalkinsd** [comment]
I created #190 to follow import optimisations, so the impact will be wider!
@Hakky54

**Comment by Hakky54** [comment]
It looks like you are enjoying all these changes 😄 
Sure I will extract the additional unit test and some code cleanup to a separate branch and scope down this branch to get more focus on the actual feature.

**Comment by Hakky54** [comment]
Any idea when you will be releasing by the way?

**Comment by rybalkinsd** [comment]
> Any idea when you will be releasing by the way?

Want to put `0.12.0` this Saturday

**Comment by Hakky54** [comment]
Hehehe, firing all the commits to make this baby ready to ship. It think I applied all your remarks. I will also push the other branches

**Pull request merged**